core/chains/evm/headtracker/heads.go

```diff
@@ -75,16 +104,20 @@ func (h *heads) AddHeads(historyDepth uint, newHeads ...*evmtypes.Head) {
 		headCopy := *head
 		headCopy.Parent = nil // always build it from scratch in case it points to a head too old to be included
 		// map eliminates duplicates
-		headsMap[head.Hash] = &headCopy
+		// prefer head that was already in heads as it might have been marked as finalized on previous run
+		if _, ok := headsMap[head.Hash]; !ok {
+			headsMap[head.Hash] = &headCopy
+		}
 	}
 
-	heads := make([]*evmtypes.Head, len(headsMap))
+	heads := make([]*evmtypes.Head, 0, len(headsMap))
 	// unsorted unique heads
 	{
-		var i int
 		for _, head := range headsMap {
-			heads[i] = head
-			i++
+			if head.BlockNumber() < minBlockToKeep {
```

Comment:
I think instead of checking twice here and during parent assignment, it's better if you add the check once at the top, when creating the map. This way you don't have to check again.

Reply:
@dhaidashenko corner case: when you assign parents down below, you search for parents in the map and not in the slice, meaning you'll always assign a parent to the finalized block, so your last block won't be the finalized one but its parent. I'm not sure if this violates any assumption, but something to be aware of.